filterEvents option control for subscribe()

iamtrep opened the pull request:
add an option to control whether attribute events where the value is not changed are filtered or not.  Default = not filtered (possibly a breaking change, but I think it's the right choice in general)

Necessary to not filter events with no value change for devices where the last value represents the average of the time period since the previous value.  This is necessary for correct "step-back interpolation", e.g. flow sensors.

iamtrep committed:
Update influxdb-logger.groovy

add an option to control whether attribute events where the value is not changed are filtered or not.  Default = not filtered (possibly a breaking change, but I think it's the right choice in general)

Necessary to not filter events with no value change for devices where the last value represents the average of the time period since the previous value.  This is necessary for correct "step-back interpolation", e.g. flow sensors.

dennypage commented:
The decision of whether or not something is considered an event is left to the driver. If the driver considers it an event, shouldn't we? What is the circumstance in which we want to override the driver's decision? A lot of people, myself included, depend upon this behavior for failsafe detection.
Also, doesn't this break the concept of soft polling?

iamtrep commented:
The driver considers it an event.  The event exists and is sent by the device (e.g. the event description text appears in the hub logs), but is filtered by Hubitat's pubsub machinery when there is no change in the attribute value since the last event sent for this attribute (see doc for subscribe() here https://docs2.hubitat.com/en/developer/app/app-object).

However the default behaviour for subscribe() is to have those events filtered, so perhaps InfluxDB-Logger should honor that default.   Seems wrong to me, but it's not something I'd argue over (and it removes the potential of a breaking change).    If you want more context for where I'm coming from, see this discussion : https://community.hubitat.com/t/sinope-water-valve-with-flow-sensor-fs4220-fs4221/96701/64)

BTW I don't understand how this could break soft polling.

iamtrep committed:
Update influxdb-logger.groovy

make filterEvents default to true, which is the default for subscribe().  Avoid breaking change w.r.t. current implementation of InfluxDB-Logger.

dennypage commented:
Hmm... Interesting. I wasn't aware of the filterEvents option. However, I think perhaps this option actually performs an override of the driver decision. 

I have physical devices which are configured to send an update every N minutes. I have soft polling disabled. I have data in influxDB wherein the same value was inserted several times by the existing version of influxdb-logger. Initially this was not the case, and a new value was being inserted only when there was a change. The reason for this was that the driver was not explicitly setting isStateChange to true in order to force creation of an event. With the driver modified to explicitly set isStateChange to true, an event is created for each physical update provided by the device, and influxdb-logger sees an update on the subscription regardless of whether or not the value has changed. I use this behavior for failsafe detection from the physical device all the way to InfluxDB.

You can also use soft polling to get around this issue, but that limits the failsafe detection to the Hubitat to InfluxDB connection (excludes the physical device). If the device is the kind of thing that you want failsafe on (like a water valve), I would hope that the device would send periodic updates and that the driver would explicitly set isStateChange to true, or at least have an option to do so.

iamtrep commented:
I won't bore you with all the details of this particular flow sensor, but it does not send events unless water flows and cannot be refreshed/polled.   Soft-polling will not work here, as it would duplicate the last-received-event attribute value in-between events and break volume calculations.

Thank you for explaining the "isStateChange" bit, I wasn't aware of that.  I do think that the driver for my flow sensor should set this to true for every event, but I have no control over the driver code as it is a system driver.  I'll poke the author.

So it sounds like defaulting this option to "true" (yet exposing it for use-cases like mine) is the correct thing to do.  You've changed my mind (had already committed that change btw).

dennypage commented:
If the physical device doesn't actually send an update, the driver won't be invoked. If the driver isn't invoked, there cannot be an event, whether of same value or new value. So I just don't see how the the filterEvents would have any effect here.

The only way I understand that this would have effect, is if the device is actually sending an update, and the value of the update is the same as the prior value. This will cause a new event if isStateChange is set to true, otherwise it will be discarded. But either way this is the same value as soft poll would return.

So, what am I missing?

iamtrep commented:
It has an effect.  My volume calculations are now accurate. 

I was only mentioning the fact that the device doesn't update unless there is flow to explain why soft-polling won't work.  When water flows, there are updates, many of them of the same value (e.g. constant flow) and all of them are important and should be logged.

Sounds like I'm describing an edge case but really all I am looking to do is to expose a built-in feature (filterEvents option) of subscribe() to the user of this app.

dennypage commented:
This is the common sequence of events that I am aware of:

1. The physical device sends a sensor report.
2. The receipt of the sensor report causes the device driver to be invoked.
3. The device driver takes sensor report and constructs a map for sendEvent()
4. The driver invokes sendEvent(). If map.value is different than the prior value, or if the driver has set map.isStateChange to true, then an event is generated.
5. The event causes The subscription callback in influxdb-logger is invoked with the value.

So, there has to be something different happening along this sequence than has been previously described.

First and foremost, I believe that the physical device must actually be sending a sensor report (step 1). Without this, I don’t see how the driver would be invoked. The best way to confirm this is with a sniffer, or possibly to talk with the driver developer to see if perhaps there is a time based trigger for something.

Second, I believe that the driver must not be explicitly setting isStateChange to true in the driver. If it were, there should be visible events created, even if the value did not change.

Third, I believe that setting filterEvents to false in the subscription must effectively override the driver’s decision in step 4, effectively causing the subscription to behave as if map.isStateChange were set to true for that particular subscription.

Regardless, the soft poll should retrieve the last published value by the driver, regardless of the decision in step 4 or the override in the subscription.

I understand that you see that the change for filterEvents has some effect. I am looking to understand why, and how it differs from soft polling.

iamtrep commented:
It works exactly as you describe in your steps.

The call to subscribe() in manageSubscriptions() does not specifiy the option parameter so subscribe is invoked with the default ["filterEvents" : true].   All my change does is expose that parameter in the UI so the user can have the subscribe() call explicitly pass ["filterEvents : false].  Then, those events that *are sent by the driver* (users can see them in the hub logs!) will be passed to InfluxDB-Logger.   This is correct behaviour : InfluxDB-Logger will be adding a data point to the queue for a **real** event with a **real** timestamp.   I can't conceive of any reason to forbid that.

There is no overriding of any driver's decision. The event gets published no matter what, whether isStateChanged is set to true or false.  It is up to the subscriber to decide if they want to have those set to false forwarded to them or not.   That is the purpose of the options parameter in subscribe().  InfluxDB-Logger being one such subscriber, the UI should permit the user to control this option.

One could even argue that by not exposing this option, the InfluxDB-Logger app is "making a decision" (to paraphrase) on behalf of the user that "these are not the events you're looking for".  "Hey, I can see these events in the hub logs, how come I can't forward them to InfluxDB?" would be a fair question IMO.

Soft polling doesn't fix this.  Soft polling creates _fake_ data points for events that have **never** been sent by any driver (overriding the drivers' _decision_).  In that sense, it's very different.  I'll say more about this on the forum.  

I feel like there's really nothing more I can add.  This is a really obvious change to me : exposes existing functionality, doesn't break any contract with device drivers, defaults to no impact for the user.  I understand it doesn't seem obvious to you.  At this point I can only recommend you go read the forum thread I linked to earlier where I explain what is going on with my sensor in detail.

dennypage commented:
Okay, do you see my confusion? Perviously you said that the device did not update, which I took to mean that the device did not send any updates, so the driver was never invoked, etc. So now what I am understanding is that the device is sending updates, but that the driver is not publishing them as events because the value is the same as previous. Okay, got it.

But this still doesn't explain how soft polling fails to product the same value. When these updates are received from the device, the driver is not declaring them to be an event because the value is exactly the same as the prior value. I.E. the last known value doesn't change. Soft polling creates these fake events based upon whatever the last known value is. These two values will be the same.

To be clear, I see value to the option because it covers the driver not explicitly setting isStateChange to true. This will allow the end to end failsafe I describe above regardless of the driver's decision about what an event is.

iamtrep commented:
` So now what I am understanding is that the device is sending updates, but that the driver is not publishing them as events because the value is the same as previous. Okay, got it.`
Not quite.   The driver is **always** publishing all updates as events, even when the value is same as before, except in that case, as you explained, the "isStateChange" flag is set to false by the driver.   

`When these updates are received from the device, the driver is not declaring them to be an event because the value is exactly the same as the prior value. `
This is inaccurate.  The driver is still sending an event, just with a different value for "isStateChanged" (I got this from you).  My understanding is that Hubitat's event pubsub uses that flag to decide whether to forward that event, or not, to subscribers, depending on what the filterEvents option was set to when subscribe() was called by each individual subscriber.

`These two values will be the same.`
The values will be the same, but the timestamps will be different, and the timestamp is an important part of the event.  The timestamp matters in how data is interpreted over time (area under curve and whatnot).

`I see value to the option`
Great !  What do you say we get this one merged and continue the discussion about soft polling on the forum.

dennypage reviewed (commented):
"Filter Events (true = only when value changes)?" seems a bit Hubitat developer oriented.

Can we use a phrasing more end user oriented, like "Only log events when the value changes"?

iamtrep committed:
better-worded description for option

dennypage reviewed (approved):
Thanks
